### PR TITLE
Torsion-free groups

### PR DESCRIPTION
**Pull request by EgbertRijke**
This PR adds the notion of torsion element of a group and and the notion of torsion-free group to the library. Furthermore, it adds a short discussion on the notion of elements of finite order, which looks too strong from a constructive point of view.

### CODE REVIEW

File: src/group-theory/elements-of-finite-order-groups.lagda.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+An element `x` of a [group](group-theory.groups.md) `G` is said to be of
+**finite order** if the [subgroup](group-theory.subgroups.md) `order x` of `ℤ`
+is [generated by](group-theory.subgroups-generated-by-elements-groups.md) a
+[nonzero integer](elementary-number-theory.nonzero-integers.md).
+
+Note that the condition of being of finite order is slightly stronger than the
+condition of being a [torsion element](group-theory.torsion-elements-groups.md).
+The latter condition merely asserts that there
+[exists](foundation.existential-quantification.md) a nonzero integer in the
+ideal `order x`.
```
Comment (fredrik-bakke) [inline src/group-theory/elements-of-finite-order-groups.lagda.md:36]: Depending on what is most appropriate, either
- link to ideal
- change wording to subgroup

File: src/group-theory/torsion-elements-groups.lagda.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+```text
+  ∃ (k : nonzero-ℤ), xᵏ ＝ 1.
+```
+
+Note that the condition of being a torsion element is slightly weaker than the
+condition of being of
+[finite order](group-theory.elements-of-finite-order-groups.md). The condition
+of being a torsion element holds
+[if and only if](foundation.logical-equivalences.md) the
+[subgroup](group-theory.subgroups.md) `order x` of `ℤ` contains a nonzero
```
Comment (fredrik-bakke) [inline src/group-theory/torsion-elements-groups.lagda.md:40]:
```suggestion
[subgroup](group-theory.subgroups.md) `order x` of `ℤ` contains a [nonzero](elementary-number-theory.nonzero-integers.md)
```

File: src/group-theory/torsion-free-groups.lagda.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+element of finite [order](group-theory.orders-of-elements-groups.md) is the
+identity element. In other words, torsion-free groups are groups in which the
+condition
+
+```text
+  ∀ (k : nonzero-ℤ), xᵏ ＝ 1 → x ＝ 1
+```
+
+holds for all elements `x : G`. This condition can be formulated in several
+equivalent ways:
```
Comment (fredrik-bakke) [inline src/group-theory/torsion-free-groups.lagda.md:42]:
```suggestion
[equivalent](foundation.logical-equivalences.md) ways:
```

**Review by fredrik-bakke — APPROVED** [review]
Looks good! Since you explicitly mention it I assume it is thought out why `is-nonzero-N` is defined elsewhere.

**Comment by EgbertRijke** [comment]
> Looks good! Since you explicitly mention it I assume it is thought out why `is-nonzero-N` is defined elsewhere.

I think it is because some very elementary stuff about natural numbers is done with the notion of nonzero natural number. My other reason was that I didn't really feel like testing it out to refactor it, so I thought I'd just mention it.

File: src/group-theory/elements-of-finite-order-groups.lagda.md (hunk 1 above)
Comment (EgbertRijke) [reply src/group-theory/elements-of-finite-order-groups.lagda.md:36]: I am changing the wording.

**Comment by fredrik-bakke** [comment]
That's okay. Given the new dependency of telescopes on natural numbers, I've actually been itching to try and move some of the dependencies in `natural-numbers`, but that's for another PR.

**Comment by EgbertRijke** [comment]
> That's okay. Given the new dependency of telescopes on natural numbers, I've actually been itching to try and move some of the dependencies in `natural-numbers`, but that's for another PR.

Do you mean to put some stuff from other files into `natural-numbers`?

**Comment by fredrik-bakke** [comment]
No, I mean the other way around. When I was playing around with telescopes, I frequently encountered cyclic dependencies because of all the things that are imported in `elementary-number-theory.natural-numbers`

**Comment by EgbertRijke** [comment]
> No, I mean the other way around. When I was playing around with telescopes, I frequently encountered cyclic dependencies because of all the things that are imported in `elementary-number-theory.natural-numbers`

That sounds good! Cyclic dependencies are so annoying!

**Comment by EgbertRijke** [comment]
TBH I find the cyclic dependencies quite dumb. A proof assistant should just test for _entries_ whether they create cyclic dependencies, not for files. Then we would truly be able to get rid of foundation-core.

**Comment by EgbertRijke** [comment]
I addressed all your comments, so it is ready

**Pull request merged**